### DISCUSSION TITLE
Is java.library.path respected?

### DISCUSSION THREAD

**Discussion by StefanOltmann** [lopcode/vips-ffm#118] [Q&A] [answered]
I want to use your lib in a [Compose Multiplatform](https://github.com/JetBrains/compose-multiplatform) Desktop app.

Adding custom DLL files to the `app` directory (part of the path) is difficult, but I can add them to a `app/resources` directory.

I get a system property where this is located:
`val resourcesDir = System.getProperty("compose.application.resources.dir")`

I try to modify `java.library.path` to get the lib loaded, but this seems not to work.

```
        System.setProperty(
            "java.library.path",
            System.getProperty("java.library.path") + ";$resourcesDir"
        )
```

Do you have any idea how I can make this work here?

**Comment by lopcode** [comment, marked as answer]
I haven't tried loading from `java.library.path`, but I would expect it to work - here are some things that might help debug:
* The code that sets where vips-ffm looks for symbols: https://github.com/lopcode/vips-ffm/blob/main/core/src/main/java/app/photofox/vipsffm/jextract/VipsRaw.java#L59 - first, it tries the system path, then falls back to `SymbolLookup.loaderLookup()` which I would expect to be able to load things from the classpath
* There are three DLLs you'll need on the path - libvips, glib, and gobject
* Every operating system has different opinions about what libraries should be named - this is the GitHub Action that checks the Windows build, you could double check the filenames are identical to a "known good" build using the system path: https://github.com/lopcode/vips-ffm/blob/main/.github/workflows/checks.yml#L78

It could be that we need to modify the loader in `VipsRaw` to allow `java.library.path` fallback - I'm open to a Pull Request if you figure out that's what's needed.

Good luck, let me know how you get on!

**Reply by StefanOltmann** [reply]
Thank you for your help.

Not the best solution, but this seems to work on Windows and is for now good enough:

```
if (PLATFORM == Platform.WINDOWS) {
  System.load("$resourcesDir/libglib-2.0-0.dll")
  System.load("$resourcesDir/libgobject-2.0-0.dll")
  System.load("$resourcesDir/libvips-42.dll")
}
```

**Reply by StefanOltmann** [reply]
Unfortunately the same trick does not work for macOS.

I pulled the libraries out of `/opt/homebrew/lib` and they have different names than you expect in `VipsLibLookup`... So I wonder why this even works.

```
        if (PLATFORM == Platform.WINDOWS) {

            System.load("$resourcesDir/libglib-2.0-0.dll")
            System.load("$resourcesDir/libgobject-2.0-0.dll")
            System.load("$resourcesDir/libvips-42.dll")
        }

        if (PLATFORM == Platform.MACOS) {

            System.load("$resourcesDir/libglib-2.0.0.dylib")
            System.load("$resourcesDir/libgobject-2.0.0.dylib")
            System.load("$resourcesDir/libvips.42.dylib")
        }

        Vips.init()
```

**Reply by lopcode** [reply]
Did you get anywhere with macOS? I'm curious if there was a loading error shown, or if the aymbol lookups just failed even though you loaded the libraries manually.

**Reply by StefanOltmann** [reply]
Nope, unfortunately not.
If you load them directly by their absolute path without having them in the system path, does it work for you?

I guess it may be related to the issue that they are named differently than Windows & Linux and you don’t check for those names.

**Reply by StefanOltmann** [reply]
@lopcode So you have any idea here?

Renaming the dynlib files also didn’t work.

what else can I try?

**Reply by lopcode** [reply]
I'll try to reproduce this at the weekend 🤞

**Reply by lopcode** [reply]
Had a look this evening, and decided it would be good to have a "first party" way of manually specifying library paths. `1.3.0` contains support for three new system properties to override a library path that should help you, and the sample runner includes a test to make sure it works.

Properties at the bottom of the "native library" section of the README: https://github.com/lopcode/vips-ffm?tab=readme-ov-file#native-library-loading
And a macOS sample for libvips in the runner here (but you'll need to set the other two as well for your use case): https://github.com/lopcode/vips-ffm/blob/main/run_samples.sh#L11

**Reply by StefanOltmann** [reply]
Thank you for the fix and your support! 😊

According to my tests VIPS is about 4x faster at generating thumbnails compared to Skiko (the SKIA wrapper)! 🚀

I'll be incorporating your library into the next release of [Ashampoo Photo Organizer](https://www.ashampoo.com/photo-organizer). 

The base app is free, with advanced features available via a subscription plan. If you're interested, I'd love to gift you a license. :)

Maybe you find some inspiration for PhotoFox. ;)

**Reply by lopcode** [reply]
Looks great! Glad it's performing well for you, that sounds similar to performance improvements others have shared too, which is nice to confirm.

And yes I'd happily accept a license to play around with the product - I made an account with the email `ashampoo`@`the website in my GitHub profile.com` 🙂

**Reply by StefanOltmann** [reply]
Yes, VIPS is truly amazing. :)

Unfortunately, the publishing team just informed me that they’re unable to sign my EXE file. The error code suggests it might be related to this issue: [JDK-8326447](https://bugs.openjdk.org/browse/JDK-8326447). I was on Java 17 before and tried to update to Java 22.

As a result, the update including VIPS and faster thumbnail generation will be delayed a bit. Quite disappointing. :(

On the bright side, I’ve just sent you your free license! :)

**Reply by lopcode** [reply]
Got it, thanks!

Hope the signing issues get resolved soon 🤞

**Reply by StefanOltmann** [reply]
In the meantime, I’m glad to have [vipsthumbnail](https://www.libvips.org/API/current/Using-vipsthumbnail.html). I can run it from the command line for now and it’s even faster than SKIA. That'll be in the next update.

The vips-ffm integration is currently on hold in a branch, ready to go live as soon as Oracle gets their act together.
